ui/components/src/utils/chart-actions.ts

```diff
@@ -147,3 +148,50 @@ export function batchDispatchNearbySeriesActions(
     });
   }
 }
+
+/*
+ * Determine whether a markLine was pushed into the final series, which means crosshair was already pinned onClick
+ */
+export function checkCrosshairPinnedStatus(seriesMapping: TimeChartSeriesMapping) {
+  const isCrosshairPinned = seriesMapping[seriesMapping.length - 1]?.name === PINNED_CROSSHAIR_SERIES_NAME;
+  return isCrosshairPinned;
+}
+
+/*
+ * Find closest timestamp to logical x coordinate returned from echartsInstance.convertFromPixel
+ * Used to find nearby series in time series tooltip.
+ */
+export function getClosestTimestamp(
+  cursorX: number,
+  currentClosestTimestamp: null | number,
```

Comment:
Cleaning this up now since currentClosestTimestamp and currentClosestDistance aren't necessary as params

Reply:
Proposing we change to below lines to improve perf, before we were looping through each inner dataset Array but this shouldn't be necessary thanks to `getCommonTimeScale`
```
// Only need to loop through first dataset source since getCommonTimeScale ensures xAxis timestamps are consistent
const firstDatasetValues = data[0]?.values;
const closestTimestamp = firstDatasetValues ? getClosestTimestamp(cursorX, firstDatasetValues) : null;
 ```

Reply:
Confirmed with Eunice offline, here is additional PR with that change in #1330